src/language-js/print/ternary.js

```diff
@@ -153,6 +158,110 @@ function printTernaryTest(path, options, print) {
   return printed;
 }
 
+const ancestorNameMap = new Map([
+  ["AssignmentExpression", "right"],
+  ["VariableDeclarator", "init"],
+  ["ReturnStatement", "argument"],
+  ["ThrowStatement", "argument"],
+  ["UnaryExpression", "argument"],
+  ["YieldExpression", "argument"],
+]);
+function shouldExtraIndentForConditionalExpression(path) {
+  const node = path.getValue();
+  if (node.type !== "ConditionalExpression") {
+    return false;
+  }
+
+  const parentName = path.callParent((path) => path.getName());
+  if (!parentName) {
+    return false;
+  }
+
+  let ancestorCount = 1;
+  let ancestor = path.getParentNode(ancestorCount);
+  while (isChainElement(ancestor)) {
+    ancestor = path.getParentNode(++ancestorCount);
+  }
+  const checkAncestor = (node) => {
+    const name = ancestorNameMap.get(ancestor.type);
+    return ancestor[name] === node;
+  };
+
+  const parent = path.getParentNode();
+
+  /**
+   * foo = (
+   *   condition
+   *     ? first
+   *     : second
+   * ) as SomeType;
+   */
+  if (parent.type === "TSAsExpression" && checkAncestor(parent)) {
+    return true;
+  }
+
+  /**
+   * foo = new (
+   *   condition
+   *     ? first
+   *     : second
+   * )(arguments);
+   */
+  if (
+    parent.type === "NewExpression" &&
+    checkAncestor(parent) &&
+    path.getName() === "callee"
+  ) {
+    return true;
+  }
+
+  const chainRoot = path.getParentNode(ancestorCount - 1);
```

Comment:
The rest part in this function seems just `CallExpression.callee`, `MemberExpression.object`, and `TSNonNullExpression.object`? Code like this should work?

```js
if (
  isChainElement(parent) &&
  checkAncestor(chainRoot) &&
  (
    (isCallExpression() && name === "callee") ||
    (MemberExpression() && name === "object") ||
    (parent.type === 'TSNonNullExpression' && name === "object") ||
  )
)
```


Note: `name === "object"` was not checking, we should add tests for them.